core/editor/src/tools/ellipse.rs

```diff
@@ -58,37 +62,112 @@ impl Fsm for EllipseToolFsmState {
 			}
 
 			(EllipseToolFsmState::LmbDown, Event::MouseMove(mouse_state)) => {
+				data.drag_current = *mouse_state;
+
 				operations.push(Operation::ClearWorkingFolder);
-				operations.push(Operation::AddCircle {
-					path: vec![],
-					insert_index: -1,
-					cx: data.drag_start.x as f64,
-					cy: data.drag_start.y as f64,
-					r: data.drag_start.distance(&mouse_state),
-					style: style::PathStyle::new(None, Some(style::Fill::new(tool_data.primary_color))),
-				});
+				operations.push(make_operation(data, tool_data));
 
 				EllipseToolFsmState::LmbDown
 			}
 
 			(EllipseToolFsmState::LmbDown, Event::LmbUp(mouse_state)) => {
-				let r = data.drag_start.distance(&mouse_state.position);
-				log::info!("draw ellipse with radius: {:.2}", r);
+				data.drag_current = mouse_state.position;
+
 				operations.push(Operation::ClearWorkingFolder);
-				operations.push(Operation::AddCircle {
-					path: vec![],
-					insert_index: -1,
-					cx: data.drag_start.x as f64,
-					cy: data.drag_start.y as f64,
-					r: data.drag_start.distance(&mouse_state.position),
-					style: style::PathStyle::new(None, Some(style::Fill::new(tool_data.primary_color))),
-				});
+				operations.push(make_operation(data, tool_data));
 				operations.push(Operation::CommitTransaction);
 
 				EllipseToolFsmState::Ready
 			}
 
+			(state, Event::KeyDown(Key::KeyShift)) => {
+				data.constrain_to_circle = true;
+
+				if state == EllipseToolFsmState::LmbDown {
+					operations.push(Operation::ClearWorkingFolder);
+					operations.push(make_operation(data, tool_data));
+				}
+
+				self
+			}
+
+			(state, Event::KeyUp(Key::KeyShift)) => {
+				data.constrain_to_circle = false;
+
+				if state == EllipseToolFsmState::LmbDown {
+					operations.push(Operation::ClearWorkingFolder);
+					operations.push(make_operation(data, tool_data));
+				}
+
+				self
+			}
+
+			(state, Event::KeyDown(Key::KeyAlt)) => {
+				data.center_around_cursor = true;
+
+				if state == EllipseToolFsmState::LmbDown {
+					operations.push(Operation::ClearWorkingFolder);
+					operations.push(make_operation(data, tool_data));
+				}
+
+				self
+			}
+
+			(state, Event::KeyUp(Key::KeyAlt)) => {
+				data.center_around_cursor = false;
+
+				if state == EllipseToolFsmState::LmbDown {
+					operations.push(Operation::ClearWorkingFolder);
+					operations.push(make_operation(data, tool_data));
+				}
+
+				self
+			}
+
 			_ => self,
 		}
 	}
 }
+
+fn make_operation(data: &EllipseToolData, tool_data: &DocumentToolData) -> Operation {
+	let x0 = data.drag_start.x as f64;
+	let y0 = data.drag_start.y as f64;
+	let x1 = data.drag_current.x as f64;
+	let y1 = data.drag_current.y as f64;
+
+	let (cx, cy) = if data.center_around_cursor { (x0, y0) } else { ((x0 + x1) * 0.5, (y0 + y1) * 0.5) };
+
+	let (mut rx, mut ry) = if data.constrain_to_circle {
+		let r = f64::max((x1 - x0).abs(), (y1 - y0).abs());
```

Comment:
maybe this should use the min instead?

Reply:
Author's choice, see my comment above.

Reply:
It's not equivalent though.`max` means the radius will be the larger of the two side lengths,
`min` means the shorter. I would think the shorter to be better since then the circle fits into the drawn rectangle and also since affinity does it like that iirc

Reply:
and I think you meant to reply to the previous (up by one) comment

Reply:
Sorry, somehow this comment ended up in the wrong thread. I was replying in reference to `/= 2.0;`

Reply:
GIMP uses max for their ellipse tool. That way, the starting and ending cursor positions are both on the bounding square of the circle. This will feel more intuitive when overlays are in place. I would however be fine with both.